quizer_game/views.py

```diff
@@ -1,3 +1,149 @@
+# TODO provide upvote-downvote feature
+# /quizer/game/player_id/quiz_id/difficulty/update/
+def update_game(request, player_id, quiz_id, selected_difficulty, choice_value):
+    quiz = get_object_or_404(Quiz, pk=quiz_id)
+    player = quiz.player_set.get(pk=player_id)
+
+    # update position
+    if choice_value == CHOICE_VALUE['correct']:
+        if player.position < POSITION['max']:
+            player.move_forward()
+    else:
+        if selected_difficulty > DIFFICULTY['easy']:
+            if player.position > POSITION['min']:
+                player.move_backward()
+
+    if player.position == POSITION['max']:
+        player.is_achieved = True
+        player.is_playing = False
+        player.save()
+        return redirect(reverse('quizer_game:result',
+                                kwargs={'player_id': player.id, 'quiz_id': quiz.id,
+                                        'selected_difficulty': player.selected_difficulty, }
+                                ),
+                        )
+
+    elif player.position < POSITION['max']:
+        # change to next question
+        try:
+            old_question = player.current_question
+            new_question_number = old_question.number + 1
+            player.current_question = quiz.question_set.get(number=new_question_number)
+        except ObjectDoesNotExist:
+            player.is_failed = True
+            player.is_playing = False
+            player.save()
+            return redirect(reverse('quizer_game:result',
+                                    kwargs={'player_id': player.id, 'quiz_id': quiz.id,
+                                            'selected_difficulty': player.selected_difficulty, }
+                                    ),
+                            )
+
+    player.save()
+    return redirect(reverse('quizer_game:game',
+                            kwargs={'player_id': player.id, 'quiz_id': quiz.id,
+                                    'selected_difficulty': selected_difficulty}
+                            )
+                    )
+
+
+# game/<int:player_id>/<int:quiz_id>/<int:selected_difficulty>/result/
+def result(request, player_id, quiz_id, selected_difficulty):
+    quiz = get_object_or_404(Quiz, pk=quiz_id)
+    player = quiz.player_set.get(pk=player_id)
+    context = {'quiz': quiz, 'player': player}
+    return render(request, 'quizer_game/result.html', context)
+
+
+def leaderboard_index(request):
+    quiz = Quiz.objects.all()
+    context = {'quizzes': quiz}
+    return render(request, 'quizer_game/leaderboard-index.html', context)
+
+
+def login(request):
+    return render(request, 'quizer_game/login.html')
+
+
+def leaderboard(request, quiz_id, selected_difficulty):
+    quiz = get_object_or_404(Quiz, pk=quiz_id)
+    player = quiz.player_set.get()
+    number = range(1, player.id+1)
+    context = {'quizzes': quiz, 'player': player, 'number': number}
+    return render(request, 'quizer_game/leaderboard.html', context)
```

Comment:
This part causes an error when there are more than 1 player played this quiz.

from `/quizer/leaderboard/4/0/`
![image](https://user-images.githubusercontent.com/48724051/69483206-dc0c6e80-0e56-11ea-8c25-bd390bffe0c6.png)

Suggest to use quiz.player_set.all() to get a query set of players instead.

Reply:
And this view doesn't only get all players to display on a template, but also rank all players. Please rank all players using their time spent then put this data to `context` to use further on `leaderboard.html`.

Reply:
Already fixed.